File: epistimi-frontend/src/dto/subject-grades.ts
import { ClassificationGradeTeacherResponse, ClassificationGradeValueResponse } from './classification-grade';
import { GradeResponse } from './grade';

export interface SubjectGradesResponse {
  id: string;
  name: string;
  students: SubjectGradesStudentResponse[];
  average: SubjectGradesAverageResponse;
}

export interface SubjectGradesStudentResponse {
  id: string;
  firstName: string;
  lastName: string;
  firstSemester: SubjectGradesStudentSemesterResponse;
  secondSemester: SubjectGradesStudentSemesterResponse;
  average?: string;
  classification: SubjectStudentClassificationResponse;
}

export interface SubjectGradesStudentSemesterResponse {
  grades: GradeResponse[];
  average?: string;
}

export interface SubjectStudentClassificationResponse {
  proposal?: SubjectStudentClassificationGradeResponse;
  final?: SubjectStudentClassificationGradeResponse;
}

export interface SubjectStudentClassificationGradeResponse {
  id: string;
  issuedBy: ClassificationGradeTeacherResponse;
  issuedAt: Date;
  updatedAt?: Date;
  value: ClassificationGradeValueResponse;
}

export interface SubjectGradesAverageResponse {
  firstSemester?: string;
  secondSemester?: string;
  overall?: string;
}

File: epistimi-backend/src/main/kotlin/pl/edu/wat/wcy/epistimi/grade/domain/SubjectGrades.kt
package pl.edu.wat.wcy.epistimi.grade.domain

import pl.edu.wat.wcy.epistimi.student.domain.StudentId
import pl.edu.wat.wcy.epistimi.subject.domain.SubjectId
import java.math.BigDecimal

data class SubjectGrades(
    val id: SubjectId,
    val name: String,
    val students: List<SubjectStudentGradesSummary>,
    val average: SubjectGradesAverageSummary,
)

data class SubjectStudentGradesSummary(
    val id: StudentId,
    val firstName: String,
    val lastName: String,
    val firstSemester: SubjectStudentSemesterGradesSummary,
    val secondSemester: SubjectStudentSemesterGradesSummary,
    val average: BigDecimal?,
)

data class SubjectStudentClassification(
    val proposal: ClassificationGrade?,
    val final: ClassificationGrade?,
)

data class SubjectStudentSemesterGradesSummary(
    val grades: List<Grade>,
    val average: BigDecimal?,
)

data class SubjectGradesAverageSummary(
    val firstSemester: BigDecimal?,
    val secondSemester: BigDecimal?,
    val overall: BigDecimal?,
)
